File: ru.kmorozov.api.onedrive/src/main/java/ru/kmorozov/onedrive/tasks/DownloadTask.java
package ru.kmorozov.onedrive.tasks;

import com.google.api.client.util.Preconditions;
import ru.kmorozov.onedrive.client.OneDriveItem;
import ru.kmorozov.onedrive.client.utils.LogUtils;
import ru.kmorozov.onedrive.client.downloader.ResumableDownloader;
import ru.kmorozov.onedrive.client.downloader.ResumableDownloaderProgressListener;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.File;
import java.io.IOException;

public class DownloadTask extends Task {

    private static final Logger log = LogManager.getLogger(DownloadTask.class.getName());
    private final File parent;
    private final OneDriveItem remoteFile;
    private final boolean replace;
    private final int chunkSize;

    public DownloadTask(Task.TaskOptions options, File parent, OneDriveItem remoteFile, boolean replace, int chunkSize) {
        super(options);

        this.parent = Preconditions.checkNotNull(parent);
        this.remoteFile = Preconditions.checkNotNull(remoteFile);
        this.replace = Preconditions.checkNotNull(replace);
        this.chunkSize = chunkSize;

        if (!parent.isDirectory()) {
            throw new IllegalArgumentException("Specified parent is not a folder");
        }
    }

    public DownloadTask(Task.TaskOptions options, File parent, OneDriveItem remoteFile, boolean replace) {
        this(options, parent, remoteFile, replace, ResumableDownloader.MAXIMUM_CHUNK_SIZE);
    }

    public int priority() {
        return 50;
    }

    @Override
    public String toString() {
        return "Download " + this.remoteFile.getFullName();
    }

    @Override
    protected void taskBody() throws IOException {

        if (Task.isIgnored(this.remoteFile)) {
            this.reporter.skipped();
            return;
        }

        if (this.remoteFile.isDirectory()) {

            File newParent = this.fileSystem.createFolder(this.parent, this.remoteFile.getName());
            this.queue.add(new UpdatePropertiesTask(this.getTaskOptions(), this.remoteFile, newParent));

            for (OneDriveItem item : this.api.getChildren(this.remoteFile)) {
                this.queue.add(new DownloadTask(this.getTaskOptions(), newParent, item, false));
            }

        } else {

            if (Task.isSizeInvalid(this.remoteFile)) {
                this.reporter.skipped();
                return;
            }

            long startTime = System.currentTimeMillis();

            File downloadFile = null;

            try {
                downloadFile = this.fileSystem.createFile(this.parent, this.remoteFile.getName() + ".tmp");

                // The progress reporter
                ResumableDownloaderProgressListener progressListener = new ResumableDownloaderProgressListener() {

                    private long startTimeInner = System.currentTimeMillis();

                    @Override
                    public void progressChanged(ResumableDownloader downloader) {

                        switch (downloader.getDownloadState()) {
                            case MEDIA_IN_PROGRESS:
                                long elapsedTimeInner = System.currentTimeMillis() - this.startTimeInner;

                                DownloadTask.this.reporter.info(String.format("Downloaded chunk (progress %.1f%%) of %s (%s/s) for file %s",
                                                            downloader.getProgress() * 100.0,
                                                            LogUtils.readableFileSize((long) downloader.getChunkSize()),
                                        0L < elapsedTimeInner ? LogUtils.readableFileSize((double) downloader.getChunkSize() / ((double) elapsedTimeInner / 1000.0d)) : 0,
                                        DownloadTask.this.remoteFile.getFullName()));

                                this.startTimeInner = System.currentTimeMillis();
                                break;
                            case MEDIA_COMPLETE:
                                long elapsedTime = System.currentTimeMillis() - startTime;
                                DownloadTask.this.reporter.info(String.format("Downloaded %s in %s (%s/s) of %s file %s",
                                                            LogUtils.readableFileSize(DownloadTask.this.remoteFile.getSize()),
                                                            LogUtils.readableTime(elapsedTime),
                                        0L < elapsedTime ? LogUtils.readableFileSize((double) DownloadTask.this.remoteFile.getSize() / ((double) elapsedTime / 1000.0d)) : 0,
                                        DownloadTask.this.replace ? "replaced" : "new",
                                        DownloadTask.this.remoteFile.getFullName()));
                        }
                    }
                };

                this.api.download(this.remoteFile, downloadFile, progressListener, this.chunkSize);

                // Do a CRC check on the downloaded file
                if (!this.fileSystem.verifyCrc(downloadFile, this.remoteFile.getCrc32())) {
                    throw new IOException(String.format("Download of file '%s' failed", this.remoteFile.getFullName()));
                }

                this.fileSystem.setAttributes(
                        downloadFile,
                        this.remoteFile.getCreatedDateTime(),
                        this.remoteFile.getLastModifiedDateTime());

                File localFile = new File(this.parent, this.remoteFile.getName());

                this.fileSystem.replaceFile(localFile, downloadFile);
                this.reporter.fileDownloaded(this.replace, this.remoteFile.getSize());
            } catch (Throwable e) {
                if (null != downloadFile) {
                    if (!downloadFile.delete()) {
                        this.reporter.warn("Unable to remove temporary file " + downloadFile.getPath());
                    }
                }

                throw e;
            }
        }
    }
}

File: ru.kmorozov.api.onedrive/src/main/java/ru/kmorozov/onedrive/tasks/DownloadTask.kt
package ru.kmorozov.onedrive.tasks

import com.google.api.client.util.Preconditions
import org.apache.logging.log4j.LogManager
import ru.kmorozov.onedrive.client.OneDriveItem
import ru.kmorozov.onedrive.client.downloader.ResumableDownloader
import ru.kmorozov.onedrive.client.downloader.ResumableDownloaderProgressListener
import ru.kmorozov.onedrive.client.utils.LogUtils
import java.io.File
import java.io.IOException

class DownloadTask @JvmOverloads constructor(options: TaskOptions, parent: File, remoteFile: OneDriveItem, replace: Boolean, private val chunkSize: Int = ResumableDownloader.MAXIMUM_CHUNK_SIZE) : Task(options) {
    private val parent: File
    private val remoteFile: OneDriveItem
    private val replace: Boolean

    init {

        this.parent = Preconditions.checkNotNull(parent)
        this.remoteFile = Preconditions.checkNotNull(remoteFile)
        this.replace = Preconditions.checkNotNull(replace)

        if (!parent.isDirectory) {
            throw IllegalArgumentException("Specified parent is not a folder")
        }
    }

    public override fun priority(): Int {
        return 50
    }

    override fun toString(): String {
        return "Download " + remoteFile.fullName
    }

    @Throws(IOException::class)
    override fun taskBody() {

        if (isIgnored(remoteFile)) {
            reporter.skipped()
            return
        }

        if (remoteFile.isDirectory) {

            val newParent = fileSystem.createFolder(parent, remoteFile.name)
            queue.add(UpdatePropertiesTask(taskOptions, remoteFile, newParent))

            for (item in api.getChildren(remoteFile)) {
                queue.add(DownloadTask(taskOptions, newParent, item, false))
            }

        } else {

            if (isSizeInvalid(remoteFile)) {
                reporter.skipped()
                return
            }

            val startTime = System.currentTimeMillis()

            var downloadFile: File? = null

            try {
                downloadFile = fileSystem.createFile(parent, remoteFile.name + ".tmp")

                // The progress reporter
                val progressListener = object : ResumableDownloaderProgressListener {

                    private var startTimeInner = System.currentTimeMillis()

                    override fun progressChanged(downloader: ResumableDownloader) {

                        when (downloader.downloadState) {
                            ResumableDownloader.DownloadState.MEDIA_IN_PROGRESS -> {
                                val elapsedTimeInner = System.currentTimeMillis() - startTimeInner

                                reporter.info(String.format("Downloaded chunk (progress %.1f%%) of %s (%s/s) for file %s",
                                        downloader.progress * 100.0,
                                        LogUtils.readableFileSize(downloader.chunkSize.toLong()),
                                        if (0L < elapsedTimeInner) LogUtils.readableFileSize(downloader.chunkSize.toDouble() / (elapsedTimeInner.toDouble() / 1000.0)) else 0,
                                        remoteFile.fullName))

                                startTimeInner = System.currentTimeMillis()
                            }
                            ResumableDownloader.DownloadState.MEDIA_COMPLETE -> {
                                val elapsedTime = System.currentTimeMillis() - startTime
                                reporter.info(String.format("Downloaded %s in %s (%s/s) of %s file %s",
                                        LogUtils.readableFileSize(remoteFile.size),
                                        LogUtils.readableTime(elapsedTime),
                                        if (0L < elapsedTime) LogUtils.readableFileSize(remoteFile.size.toDouble() / (elapsedTime.toDouble() / 1000.0)) else 0,
                                        if (replace) "replaced" else "new",
                                        remoteFile.fullName))
                            }
                            ResumableDownloader.DownloadState.NOT_STARTED -> TODO()
                        }
                    }
                }

                api.download(remoteFile, downloadFile, progressListener, chunkSize)

                // Do a CRC check on the downloaded file
                if (!fileSystem.verifyCrc(downloadFile, remoteFile.crc32)) {
                    throw IOException(String.format("Download of file '%s' failed", remoteFile.fullName))
                }

                fileSystem.setAttributes(
                        downloadFile,
                        remoteFile.createdDateTime,
                        remoteFile.lastModifiedDateTime)

                val localFile = File(parent, remoteFile.name)

                fileSystem.replaceFile(localFile, downloadFile)
                reporter.fileDownloaded(replace, remoteFile.size)
            } catch (e: Throwable) {
                if (null != downloadFile) {
                    if (!downloadFile.delete()) {
                        reporter.warn("Unable to remove temporary file " + downloadFile.path)
                    }
                }

                throw e
            }

        }
    }

    companion object {

        private val log = LogManager.getLogger(DownloadTask::class.java.name)
    }
}

